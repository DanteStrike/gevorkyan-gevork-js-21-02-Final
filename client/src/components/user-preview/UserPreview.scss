@import "src/assets/variables";

.user-preview {
  position: relative;
  display: flex;
  width: 100%;
  min-height: 395px;
  padding: 2%;
  margin-bottom: 15px;
  background-color: var(--primary-background);
  border: var(--primary-border);
  border-radius: 15px;

  &__img {
    width: 30%;
    margin-right: 20px;
    background-color: var(--body-background);
    border-radius: 15px;
    object-fit: cover;
  }

  &__info {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    width: 30%;
  }

  &__title {
    margin-bottom: 20px;
    font-weight: 700;
  }

  &__row {
    margin-bottom: 5px;
  }

  &__row--id {
    margin-top: auto;
  }

  &__prop {
    font-weight: 700;
  }

  &__edit {
    display: flex;
    align-items: center;
    align-self: flex-start;
    padding: 0;
    margin: 0 0 0 auto;
    color: var(--secondory-font-color);
  }

  &__edit:hover {
    opacity: 0.8;
  }

  &__edit:active {
    opacity: 0.6;
  }

  &__edit-text {
    margin-left: 6px;
    color: inherit;
  }
}

@media (max-width: $tabletWidth) {
  .user-preview {
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    font-size: 16px;

    &__img {
      min-width: 180px;
      min-height: 180px;
      margin-right: 0;
    }

    &__info {
      display: flex;
      align-items: center;
      width: 100%;
    }

    &__edit {
      margin: 0 auto;
    }
  }
}
