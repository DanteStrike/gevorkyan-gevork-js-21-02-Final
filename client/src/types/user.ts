export type UserTitleType = 'mr' | 'ms' | 'mrs' | 'miss' | 'dr' | '';
export type UserGenderType = 'male' | 'female' | 'other' | '';

export interface IUserPreview {
  id: string;
  title?: UserTitleType;
  firstName: string;
  lastName: string;
  picture?: string;
}

export interface ILocation {
  street: string;
  city: string;
  state: string;
  country: string;
  timezone: string;
}

export interface IUser extends IUserPreview {
  gender: UserGenderType;
  email: string;
  dateOfBirth: string;
  registerDate: string;
  phone: string;
  location: ILocation;
}

export interface IUserRegistration {
  firstName: string;
  lastName: string;
  gender: UserGenderType;
  dateOfBirth: string;
  email: string;
  phone: string;
}

export interface IUserUpdate extends Omit<IUserRegistration, 'email'> {
  id: string;
  picture: string;
}
