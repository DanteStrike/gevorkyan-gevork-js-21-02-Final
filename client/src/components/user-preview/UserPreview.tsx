import React from 'react';
import './UserPreview.scss';
import {Link} from 'react-router-dom';
import {EditOutlined} from '@ant-design/icons';
import {IUser} from '../../types';
import Loading from '../loading/Loading';
import {DataUtils, RouteUtils} from '../../utils';
import useAppTranslation from '../../hooks/use-app-translation';

interface IUserPreviewProps {
  isUser: boolean;
  user: IUser;
  isLoading: boolean;
}

function UserPreview({user, isLoading, isUser}: IUserPreviewProps) {
  const {t} = useAppTranslation(`profile`);
  const {id, picture, title, firstName, lastName, gender, email, dateOfBirth, registerDate, phone} = user;
  const name = DataUtils.collectFullName(firstName, lastName, title);
  const profileEditRoute = RouteUtils.createProfileEditRoute(id);

  return (
    <article className="user-preview">
      <Loading isLoading={isLoading} />
      {picture ? <img className="user-preview__img" alt={name} src={picture} /> : <div className="user-preview__img" />}
      <p className="user-preview__info">
        <span className="user-preview__title">{name}</span>
        <span className="user-preview__row">
          <span className="user-preview__prop">{t(`gender`)}</span> {gender}
        </span>
        <span className="user-preview__row">
          <span className="user-preview__prop">{t(`dateOfBirth`)}</span> {dateOfBirth}
        </span>
        <span className="user-preview__row">
          <span className="user-preview__prop">{t(`registerDate`)}</span> {registerDate}
        </span>
        <span className="user-preview__row">
          <span className="user-preview__prop">{t(`email`)}</span> {email}
        </span>
        <span className="user-preview__row">
          <span className="user-preview__prop">{t(`phone`)}</span> {phone}
        </span>
        <span className="user-preview__row user-preview__row--id">
          <span className="user-preview__prop">ID:</span> {id}
        </span>
      </p>
      {isUser && (
        <div className="user-preview__edit">
          <Link className="user-preview__edit-text" to={profileEditRoute}>
            <EditOutlined />
            {t(`edit`)}
          </Link>
        </div>
      )}
    </article>
  );
}

export default React.memo(UserPreview);
