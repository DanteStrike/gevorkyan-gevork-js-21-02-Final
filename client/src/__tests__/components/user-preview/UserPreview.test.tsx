import React from 'react';
import {shallow} from 'enzyme';
import UserPreview from '../../../components/user-preview/UserPreview';
import {fullUser} from '../../mocks/user.mock';

describe(`Component UserPreview should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(<UserPreview isUser={false} user={fullUser} isLoading={false} />);

    expect(wrapper.find(`.user-preview__edit`)).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  it(`snapshot auth user test`, () => {
    const wrapper = shallow(<UserPreview isUser user={fullUser} isLoading={false} />);

    expect(wrapper.find(`.user-preview__edit`)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component UserPreview should work correctly`, () => {
  it(`loading test`, () => {
    const wrapper = shallow(<UserPreview isUser={false} user={fullUser} isLoading />);

    expect(wrapper.find(`Memo(Loading)`).prop(`isLoading`)).toBe(true);
    wrapper.setProps({isLoading: false});
    expect(wrapper.find(`Memo(Loading)`).prop(`isLoading`)).toBe(false);
  });
});
