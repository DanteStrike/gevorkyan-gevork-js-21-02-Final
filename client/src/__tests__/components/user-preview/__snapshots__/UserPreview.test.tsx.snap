// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component UserPreview should render correctly snapshot auth user test 1`] = `
<article
  className="user-preview"
>
  <Memo(Loading)
    isLoading={false}
  />
  <img
    alt="Ms. antLN anyFN"
    className="user-preview__img"
    src="url"
  />
  <p
    className="user-preview__info"
  >
    <span
      className="user-preview__title"
    >
      Ms. antLN anyFN
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Gender:
      </span>
       
      female
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Date of birth:
      </span>
       
      normalized
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Register date:
      </span>
       
      normalized
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Email:
      </span>
       
      email
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Phone:
      </span>
       
      phone
    </span>
    <span
      className="user-preview__row user-preview__row--id"
    >
      <span
        className="user-preview__prop"
      >
        ID:
      </span>
       
      userID_full
    </span>
  </p>
  <div
    className="user-preview__edit"
  >
    <Link
      className="user-preview__edit-text"
      to="/profile/userID_full/edit"
    >
      <ForwardRef(EditOutlined) />
      Edit
    </Link>
  </div>
</article>
`;

exports[`Component UserPreview should render correctly snapshot default test 1`] = `
<article
  className="user-preview"
>
  <Memo(Loading)
    isLoading={false}
  />
  <img
    alt="Ms. antLN anyFN"
    className="user-preview__img"
    src="url"
  />
  <p
    className="user-preview__info"
  >
    <span
      className="user-preview__title"
    >
      Ms. antLN anyFN
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Gender:
      </span>
       
      female
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Date of birth:
      </span>
       
      normalized
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Register date:
      </span>
       
      normalized
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Email:
      </span>
       
      email
    </span>
    <span
      className="user-preview__row"
    >
      <span
        className="user-preview__prop"
      >
        Phone:
      </span>
       
      phone
    </span>
    <span
      className="user-preview__row user-preview__row--id"
    >
      <span
        className="user-preview__prop"
      >
        ID:
      </span>
       
      userID_full
    </span>
  </p>
</article>
`;
